### PR TITLE
API Skeleton 2021-01-31 - swagger

### PR DESCRIPTION
**Pull request by mjudeikis**
This is  just a skeleton builds on top of https://github.com/Azure/ARO-RP/pull/1191 

This PR does:
* enables us to generate swagger for multiple apis in more repeatable way based on features apis delivers
* Adds placeholders for some new api fields we agreed to add (SDN provider, kubeconfig download endpoint, encryption at host, more VM types (enum adding)). No feature behind them yet. 
* features will be implemented later

### CODE REVIEW

File: pkg/api/v20210131preview/openshiftclustercredentials.go [hunk 1]
```diff
@@ -11,3 +11,9 @@ type OpenShiftClusterCredentials struct {
 	// The password for the kubeadmin user.
 	KubeadminPassword string `json:"kubeadminPassword,omitempty"`
 }
+
+// OpenShiftClusterAdminCredentials represents an OpenShift cluster's admin kube-config credentials
+type OpenShiftClusterAdminCredentials struct {
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftclustercredentials.go:16]: new file for this

File: pkg/api/v20210131preview/openshiftclustercredentials.go [hunk 2]
```diff
@@ -11,3 +11,9 @@ type OpenShiftClusterCredentials struct {
 	// The password for the kubeadmin user.
 	KubeadminPassword string `json:"kubeadminPassword,omitempty"`
 }
+
+// OpenShiftClusterAdminCredentials represents an OpenShift cluster's admin kube-config credentials
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftclustercredentials.go:15]: OpenShiftClusterAdminKubeconfig?

File: pkg/api/v20210131preview/openshiftclustercredentials.go [hunk 3]
```diff
@@ -11,3 +11,9 @@ type OpenShiftClusterCredentials struct {
...
+type OpenShiftClusterAdminCredentials struct {
+	// KubeConfig - Base64-encoded Kubernetes configuration file
+	KubeConfig []byte `json:"kubeConfig,omitempty"`
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftclustercredentials.go:18]: I still prefer `json:"kubeconfig,omitempty"` ;-)

**Review by jim-minter — COMMENTED** [review]
This is looking pretty plausible to me.  Let's catch up on this tomorrow.

File: pkg/api/v20210131preview/openshiftadmincredentials.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package v20210131preview
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// OpenShiftClusterAdminKubeconfig represents an OpenShift cluster's admin kube-config credentials
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftadmincredentials.go:6]:
```suggestion
// OpenShiftClusterAdminKubeconfig represents an OpenShift cluster's admin kubeconfig.
```

File: pkg/api/v20210131preview/openshiftadmincredentials.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package v20210131preview
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftadmincredentials.go:1]: rename this file to openshiftclusteradminkubeconfig.go

File: pkg/api/v20210131preview/openshiftadmincredentials.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// OpenShiftClusterAdminKubeconfig represents an OpenShift cluster's admin kube-config credentials
+type OpenShiftClusterAdminKubeconfig struct {
+	// Kubeconfig - Base64-encoded Kubernetes configuration file
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftadmincredentials.go:8]:
```suggestion
	// The base64-encoded kubeconfig file.
```

File: pkg/api/v20210131preview/openshiftadmincredentials_example.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package v20210131preview
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftadmincredentials_example.go:1]: rename this file to openshiftclusteradminkubeconfig_example.go

File: pkg/api/v20210131preview/openshiftadmincredentials_example.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// ExampleOpenShiftClusterAdminCredentialsResponse returns an example
+// OpenShiftClusterAdminCredentials object that the RP might return to an end-user
+func ExampleOpenShiftClusterAdminKubeconfigResponse() *OpenShiftClusterAdminKubeconfig {
+	return &OpenShiftClusterAdminKubeconfig{
+		Kubeconfig: []byte("base64kubeconfig"),
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftadmincredentials_example.go:10]:
```suggestion
		Kubeconfig: []byte("{}"),
```

File: pkg/api/v20210131preview/openshiftadmincredentials_example.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package v20210131preview
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// ExampleOpenShiftClusterAdminCredentialsResponse returns an example
+// OpenShiftClusterAdminCredentials object that the RP might return to an end-user
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftadmincredentials_example.go:7]:
```suggestion
// OpenShiftClusterAdminKubeconfig object that the RP might return to an end-user
```

File: swagger/redhatopenshift/resource-manager/readme.md [hunk 10]
```diff
@@ -1,7 +1,6 @@
 # Azure Red Hat OpenShift
```
Comment (jim-minter) [inline swagger/redhatopenshift/resource-manager/readme.md:1]: revert the space changes in this file

File: pkg/swagger/swagger.go [hunk 11]
```diff
@@ -73,16 +88,19 @@ func Run(api, outputDir string) error {
 
 	xmsEnumList := map[string]struct{}{}
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:89]: I think xmsEnumList should include SDNProvider

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 12]
```diff
@@ -109,22 +109,47 @@ type ServicePrincipalProfile struct {
 	ClientSecret string `json:"clientSecret,omitempty"`
 }
 
+// SDNProvider constants
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:112]:
```suggestion
// SDNProvider constants.
```

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 13]
```diff
@@ -109,22 +109,47 @@ type ServicePrincipalProfile struct {
...
+
 // NetworkProfile represents a network profile.
 type NetworkProfile struct {
 	// The CIDR used for OpenShift/Kubernetes Pods.
 	PodCIDR string `json:"podCidr,omitempty"`
 
 	// The CIDR used for OpenShift/Kubernetes Services.
 	ServiceCIDR string `json:"serviceCidr,omitempty"`
+
+	// SDNPlugin defines SDN plugin, used in the cluster.
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:128]:
```suggestion
	// The SDN plugin used in the cluster.
```

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 14]
```diff
@@ -109,22 +109,47 @@ type ServicePrincipalProfile struct {
...
+	SDNProvider SDNProvider `json:"sdnProvider,omitempty"`
 }
 
+// EncryptionAtHost constants
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:132]:
```suggestion
// EncryptionAtHost constants.
```

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 15]
```diff
@@ -109,22 +109,47 @@ type ServicePrincipalProfile struct {
...
+
 // MasterProfile represents a master profile.
 type MasterProfile struct {
 	// The size of the master VMs.
 	VMSize VMSize `json:"vmSize,omitempty"`
 
 	// The Azure resource ID of the master subnet.
 	SubnetID string `json:"subnetId,omitempty"`
+
+	// EncryptionAtHost defines value encryptionAtHost option for all master virtual machines.
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:148]:
```suggestion
	// Whether master virtual machines are encrypted at host.
```

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 16]
```diff
@@ -109,22 +109,47 @@ type ServicePrincipalProfile struct {
...
+	EncryptionAtHost EncryptionAtHost `json:"encryptionAtHost,omitempty"`
+
+	// DiskEncryptionSetID defines resourceID for diskEncryptionSet resource. It must be in the same subscription.
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:151]:
```suggestion
	// The resource ID of an associated DiskEncryptionSet, if applicable.
```

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 17]
```diff
@@ -171,6 +196,12 @@ type WorkerProfile struct {
 
 	// The number of worker VMs.
 	Count int `json:"count,omitempty"`
+
+	// EncryptionAtHost defines value encryptionAtHost option for worker profile virtual machines.
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:200]:
```suggestion
	// Whether worker virtual machines are encrypted at host.
```

File: pkg/api/v20210131preview/openshiftcluster.go [hunk 18]
```diff
@@ -171,6 +196,12 @@ type WorkerProfile struct {
...
+	EncryptionAtHost EncryptionAtHost `json:"encryptionAtHost,omitempty"`
+
+	// DiskEncryptionSetID defines resourceID for diskEncryptionSet resource. It must be in the same subscription.
```
Comment (jim-minter) [inline pkg/api/v20210131preview/openshiftcluster.go:203]:
```suggestion
	// The resource ID of an associated DiskEncryptionSet, if applicable.
```

File: pkg/swagger/swagger.go [hunk 19]
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/Azure/ARO-RP/pkg/util/stringutils"
 )
 
+var stableAPI = "github.com/Azure/ARO-RP/pkg/api/v20200430"
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:15]:
```suggestion
const stableAPI = "github.com/Azure/ARO-RP/pkg/api/v20200430"
```

Longer term we'll need to handle multiple stable versions, so this is not future-proof enough, but I guess it'll do for now.

File: pkg/swagger/swagger.go [hunk 20]
```diff
@@ -54,6 +56,19 @@ func Run(api, outputDir string) error {
...
 		},
 	}
+	// only preview api
+	if stableAPI != api {
+		s.Paths["/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openShiftClusters/{resourceName}/listAdminCredentials"] = &PathItem{
+			Post: &Operation{
+				Tags:        []string{"OpenShiftClusters"},
+				Summary:     "Lists admin credentials of an OpenShift cluster with the specified subscription, resource group and resource name.",
+				Description: "The operation returns the admin credentials.",
+				OperationID: "OpenShiftClusters_ListAdminCredentials",
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:66]:
```suggestion
				Summary:     "Lists admin kubeconfig of an OpenShift cluster with the specified subscription, resource group and resource name.",
				Description: "The operation returns the admin kubeconfig.",
				OperationID: "OpenShiftClusters_ListAdminKubeconfig",
```

File: pkg/swagger/swagger.go [hunk 21]
```diff
@@ -134,7 +152,6 @@ func Run(api, outputDir string) error {
 	}
 
 	b = append(b, '\n')
-
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:137]: don't drop the space

File: pkg/swagger/typewalker.go [hunk 22]
```diff
@@ -183,7 +183,7 @@ func (tw *typeWalker) define(definitions Definitions, name string) {
 }
 
 // define adds a Definition for the named types in the given package
-func define(definitions Definitions, pkgname string, xmsEnumList map[string]struct{}, names ...string) error {
+func define(definitions Definitions, pkgname string, xmsEnumList map[string]struct{}, names []string) error {
```
Comment (jim-minter) [inline pkg/swagger/typewalker.go:186]: could keep this the same and write `err := define(s.Definitions, api, xmsEnumList, names...)` above

File: pkg/util/azureclient/mgmt/redhatopenshift/2021-01-31-preview/redhatopenshift/operations.go [hunk 23]
```diff
@@ -41,8 +41,8 @@ func NewOperationsClient(environment *azure.Environment, subscriptionID string,
 		client = redhatopenshift.NewOperationsClientWithBaseURI(environment.ResourceManagerEndpoint, subscriptionID)
 		client.Authorizer = authorizer
 	}
-
 	return &operationsClient{
 		OperationsClient: client,
 	}
+
```
Comment (jim-minter) [inline pkg/util/azureclient/mgmt/redhatopenshift/2021-01-31-preview/redhatopenshift/operations.go:47]: more unnecessary whitespace change

**Review by jim-minter — CHANGES_REQUESTED** [review]
just a lot of nits - please fix out and we'll merge.

Converters and changes to internal API are missing here - OK with those being in a follow-up PR.

**Comment by mjudeikis** [comment]
https://github.com/Azure/azure-rest-api-specs/pull/12144 upstream review

**Comment by mjudeikis** [comment]
We had an issue with systemData.
https://github.com/Azure/azure-rest-api-specs/pull/12144/checks?check_run_id=1564328520
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources 

I interpreted this as a mandatory field on the main object we return. Hence adding it to the `OpenShiftCluster` object only.

In addition, added `struct` wrapper to manage API features in all the generators.

**Comment by mjudeikis** [comment]
@jim-minter, the last commit will require your attention.

File: pkg/swagger/swagger.go [hunk 24]
```diff
@@ -5,14 +5,49 @@ package swagger
...
+}
+
+type apiFeatures struct {
+	kubeconfig         bool
+	xmsEnum            bool
+	commonTypesVersion string
+	systemData         bool
+}
+
+var apis = map[string]apiFeatures{
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:30]: I almost wonder if this ends up going into the package itself

File: pkg/swagger/generator.go [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+var _ generator = &apiv20200430{}
+
+type features struct {
+	kubeconfig         bool
+	xmsEnum            bool
+	commonTypesVersion string
+	systemData         bool
+}
+
+type apiv20210131preview struct {
```
Comment (jim-minter) [inline pkg/swagger/generator.go:39]: I think this is pretty over engineered - you could probably simplify and get away with a static configuration like this:
```golang
type api struct {
	exampleOpenShiftClusterPutParameter func() interface{}
	exampleOpenShiftClusterPatchParameter func() interface{}
	exampleOpenShiftClusterResponse func() interface{}
	exampleOpenShiftClusterCredentialsResponse func() interface{}
	exampleOpenShiftClusterAdminKubeconfigResponse func() interface{}
	exampleOpenShiftClusterListResponse func() interface{}
	exampleOperationListResponse func() interface{}

	xmsEnum            []string
	commonTypesVersion string
	systemData         bool
}

var apis = map[string]*api {
	apiv20200430Path: {
		exampleOpenShiftClusterPutParameter: v20200430.ExampleOpenShiftClusterPutParameter,
		// ...
	},
	// ...
}

File: pkg/swagger/swagger.go [hunk 26]
```diff
@@ -12,7 +12,15 @@ import (
 	"github.com/Azure/ARO-RP/pkg/util/stringutils"
 )
 
+var g generator
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:15]: Please don't make this a global.

File: pkg/swagger/generator.go [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+
+const apiv20200430Path = "github.com/Azure/ARO-RP/pkg/api/v20200430"
+const apiv20210131previewPath = "github.com/Azure/ARO-RP/pkg/api/v20210131preview"
+
+var _ generator = &apiv20210131preview{}
+var _ generator = &apiv20200430{}
+
+type features struct {
+	kubeconfig         bool
+	xmsEnum            bool
```
Comment (jim-minter) [inline pkg/swagger/generator.go:34]: should this be a []string?

File: pkg/swagger/swagger.go [hunk 28]
```diff
@@ -73,16 +93,18 @@ func Run(api, outputDir string) error {
 
 	xmsEnumList := map[string]struct{}{}
 
-	if api != "github.com/Azure/ARO-RP/pkg/api/v20200430" {
+	names := []string{"OpenShiftClusterList", "OpenShiftClusterCredentials"}
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:96]: I think a lot of the code around here could be more neatly parameterised and defined in g

File: swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/examples/OpenShiftClusters_CreateOrUpdate.json [hunk 29]
```diff
@@ -5,7 +5,6 @@
     "resourceGroupName": "resourceGroup",
     "resourceName": "resourceName",
     "parameters": {
-      "location": "location",
```
Comment (jim-minter) [inline swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/examples/OpenShiftClusters_CreateOrUpdate.json:8]: why has this changed? x2

File: pkg/swagger/swagger.go [hunk 30]
```diff
@@ -126,6 +148,19 @@ func Run(api, outputDir string) error {
 		}
 
 		s.Definitions[azureResource].Properties = properties
+
+		if g.Features().systemData {
```
Comment (jim-minter) [inline pkg/swagger/swagger.go:152]: this doesn't look right.  I think it is masking the fact that we actually need to add this to the model and deal with it.  See https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources

File: pkg/swagger/types.go [hunk 31]
```diff
@@ -164,6 +164,7 @@ type Response struct {
 	Schema      *Schema `json:"schema,omitempty"`
 	Headers     Headers `json:"headers,omitempty"`
 	Examples    Example `json:"examples,omitempty"`
+	SystemData  *Schema `json:"systemData,omitempty"`
```
Comment (jim-minter) [inline pkg/swagger/types.go:167]: what is this for?

**Review by jim-minter — CHANGES_REQUESTED** [review]

**Comment by jim-minter** [comment]
bump - there's a longstanding unactioned review here

File: pkg/swagger/swagger.go (hunk 30 above)
Comment (jim-minter) [reply pkg/swagger/swagger.go:152]: And https://armwiki.azurewebsites.net/api_contracts/ResourceSystemData.html

File: pkg/swagger/generator.go [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+const apiv20200430Path = "github.com/Azure/ARO-RP/pkg/api/v20200430"
+const apiv20210131previewPath = "github.com/Azure/ARO-RP/pkg/api/v20210131preview"
+
+var _ generator = &apiv20210131preview{}
+var _ generator = &apiv20200430{}
+
+type features struct {
+	kubeconfig         bool
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/swagger/generator.go:33]: Easier to read and easier to understand. As much as it is possible I think consistency is better if we keep same behavior (feature flags). Otherwise some of them will be "flags, some function checks". Unreadable and insiders knowledge.

File: pkg/swagger/generator.go (hunk 25 above)
Comment (mjudeikis) [reply pkg/swagger/generator.go:39]: Can we do this in go?
```
cannot use v20200430.ExampleOpenShiftClusterPutParameter (value of type func() *v20200430.OpenShiftCluster) as func() interface{} value in struct literal
```

was sitting on this for 2 hours trying to understand what you had in mind here....
```
foo func() interface{}
!=
func foo() *Custom{
```
https://play.golang.org/p/jUb0aqTLP5s

Comment (mjudeikis) [reply pkg/swagger/generator.go:39]: Before I start refactoring for 4th time I would like to understand if Im missing something here for this to work (?)

File: pkg/swagger/types.go (hunk 31 above)
Comment (mjudeikis) [reply pkg/swagger/types.go:167]: ARM CI/sign-off required this for all new APIs/API Versions. I think we will need to make change not to log this in any away in our code but overall we need it now

File: pkg/swagger/generator.go (hunk 25 above)
Comment (gvanderpotte) [reply pkg/swagger/generator.go:39]: This caught my curiosity and found this interesting answer.
https://stackoverflow.com/a/54751503
I am still trying to digest all of it

Comment (jim-minter) [reply pkg/swagger/generator.go:39]: @mjudeikis I'm inclined to do the horrible thing and make all the Example*() functions return interface{}.  It's rubbish but it's a lot less boilerplate.

File: pkg/swagger/swagger.go [hunk 33]
```diff
@@ -126,6 +142,19 @@ func Run(api, outputDir string) error {
 		}
 
 		s.Definitions[azureResource].Properties = properties
+
+		if g.systemData {
+			s.Definitions[azureResource].Properties = append(s.Definitions[azureResource].Properties,
```
Comment (mjudeikis) [inline pkg/swagger/swagger.go:147]: wondering if this is right?

File: pkg/swagger/swagger.go (hunk 30 above)
Comment (mjudeikis) [reply pkg/swagger/swagger.go]: Added separate PR for this. Sill in draft, but working on it https://github.com/Azure/ARO-RP/pull/1412

File: pkg/swagger/swagger.go [hunk 34]
```diff
@@ -24,7 +29,7 @@ func Run(api, outputDir string) error {
 		Schemes:     []string{"https"},
 		Consumes:    []string{"application/json"},
 		Produces:    []string{"application/json"},
-		Paths:       populateTopLevelPaths("Microsoft.RedHatOpenShift", "openShiftCluster", "OpenShift cluster"),
+		Paths:       g.populateTopLevelPaths("Microsoft.RedHatOpenShift", "openShiftCluster", "OpenShift cluster"),
```
Comment (m1kola) [inline pkg/swagger/swagger.go:32]: Can we put `resourceProviderNamespace` as a parameter into `generator` struct so we don't have to pass it here?

File: pkg/swagger/swagger.go [hunk 35]
```diff
@@ -50,19 +55,31 @@ func Run(api, outputDir string) error {
...
+				Summary:     "Lists admin kubeconfig of an OpenShift cluster with the specified subscription, resource group and resource name.",
+				Description: "The operation returns the admin kubeconfig.",
+				OperationID: "OpenShiftClusters_ListAdminKubeconfig",
+				Parameters:  g.populateParameters(3, "OpenShiftCluster", "OpenShift cluster"),
+				Responses:   g.populateResponses("OpenShiftClusterAdminKubeconfig", false, http.StatusOK),
+			},
+		}
+	}
 
 	s.Paths["/providers/Microsoft.RedHatOpenShift/operations"] = &PathItem{
```
Comment (m1kola) [inline pkg/swagger/swagger.go:75]: I think we need to move `listCredentials`, `listAdminCredentials` (if we actually need it, see the question above) and `operations` paths somewhere into `generator`. Similarly to `populateTopLevelPaths`.

And probably all logic from `Run` into generator so the whole `Run` function will be something like:

```go
g, err := New(api, params1, param2)
if err != nil {
	return err
}

s := g.Swagger()

b, err := json.MarshalIndent(s, "", "  ")
if err != nil {
	return err
}

b = append(b, '\n')

return ioutil.WriteFile(outputDir+"/redhatopenshift.json", b, 0666)
```

 But I think it should be a separate PR: this one is already too big.

**Review by m1kola — COMMENTED** [review]
I like how you introduced support for generating multiple api versions and their examples, but I'm not sure that changes related to `systemData` and admin kube config are correct. For instance this PR introduces a new endpoint to list the admin kube config. Are we ready to make this decision?

I'm up for merging the generator part, but not sure that I'm comfortable with placeholders for new features and related swagger generation changes. If you want - we can split this into multiple pull requests so we can merge the generator part faster.

**Comment by mjudeikis** [comment]
> I like how you introduced support for generating multiple api versions and their examples, but I'm not sure that changes related to `systemData` and admin kube config are correct. For instance this PR introduces a new endpoint to list the admin kube config. Are we ready to make this decision?
> 
> I'm up for merging the generator part, but not sure that I'm comfortable with placeholders for new features and related swagger generation changes. If you want - we can split this into multiple pull requests so we can merge the generator part faster.

I think I gonna split the PRs as you suggest. Initially placeholder where there to kick start arm review process on API so could do implementation in parallel. We still need to do this, but maybe separate PR.

File: pkg/swagger/swagger.go [hunk 36]
```diff
@@ -50,19 +55,31 @@ func Run(api, outputDir string) error {
...
 			Description: "The operation returns the credentials.",
 			OperationID: "OpenShiftClusters_ListCredentials",
-			Parameters:  populateParameters(3, "OpenShiftCluster", "OpenShift cluster"),
-			Responses:   populateResponses("OpenShiftClusterCredentials", false, http.StatusOK),
+			Parameters:  g.populateParameters(3, "OpenShiftCluster", "OpenShift cluster"),
+			Responses:   g.populateResponses("OpenShiftClusterCredentials", false, http.StatusOK),
 		},
 	}
+	if g.kubeconfig {
+		s.Paths["/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openShiftClusters/{resourceName}/listAdminCredentials"] = &PathItem{
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/swagger/swagger.go:63]: Yes, this will be different endpoint and API.  Few reasons for this:
1. Prevent api breakage for old API
2. API purpose is different as it will download file, and not expose
3. Easier to implement. 
4. It will return privileged kubeconfig so in the future dedicated rbac could be added

**Comment by mjudeikis** [comment]
will re-open in smallers prs

File: pkg/swagger/swagger.go (hunk 34 above)
Comment (mjudeikis) [reply pkg/swagger/swagger.go:32]: Not on this PR, or one if follow-up. Too much change.
